presto-main/src/main/java/com/facebook/presto/metadata/InMemoryNodeManager.java

```diff
@@ -23,17 +23,26 @@
 import com.google.common.collect.Multimaps;
 import com.google.common.collect.SetMultimap;
 
+import javax.annotation.concurrent.GuardedBy;
 import javax.inject.Inject;
 
 import java.net.URI;
+import java.util.List;
 import java.util.Set;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.function.Consumer;
+
+import static java.util.Objects.requireNonNull;
 
 public class InMemoryNodeManager
         implements InternalNodeManager
 {
     private final Node localNode;
     private final SetMultimap<ConnectorId, Node> remoteNodes = Multimaps.synchronizedSetMultimap(HashMultimap.create());
 
+    @GuardedBy("this")
+    private final List<Consumer<AllNodes>> listeners = new CopyOnWriteArrayList<>();
```

Comment:
Purely out of curiosity - why CopyOnWriteArrayList?

Reply:
That is a mistake. Originally, I was going to use a concurrent data structure, but then decided to use a simpler synchronized design.